From 35823afc8a218f1fb778a2714400c1fa0b481fa0 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Thu, 18 Jul 2024 19:28:28 +0200
Subject: [PATCH] Add CountScriptSigOpsP2SH to count sigops in P2SH (#22)

Takes a scriptSig and counts the sigops found on the redeemScript, i.e. the last pushop interpreted as a Script.

This is based on the `CScript::GetSigOpCount` deleted in https://reviews.bitcoinabc.org/D6105; but here it's assumed the script is P2SH, which simplifies the implementation.

(cherry picked from commit 33928d4ea8549ff331b99f0861e703c267417a46)

diff --git a/src/script/sigops.cpp b/src/script/sigops.cpp
index a995342665f916f59f85a2727bb31060d856009b..bcfa0a829800622c26d3c0426dc4d3b3fd39a2cf 100644
--- a/src/script/sigops.cpp
+++ b/src/script/sigops.cpp
@@ -49,3 +49,22 @@ uint32_t CountScriptSigOps(const CScript &script, SigOpCountMode mode) {
 
     return nSigOps;
 }
+
+uint32_t CountScriptSigOpsP2SH(const CScript &scriptSig) {
+    // Get the last item that the scriptSig pushes onto the stack:
+    CScript::const_iterator pc = scriptSig.begin();
+    std::vector<uint8_t> vData;
+    while (pc < scriptSig.end()) {
+        opcodetype opcode;
+        if (!scriptSig.GetOp(pc, opcode, vData)) {
+            return 0;
+        }
+        if (opcode > OP_16) {
+            return 0;
+        }
+    }
+
+    // ... and return its opcount, using "ACCURATE" counting:
+    CScript subscript(vData.begin(), vData.end());
+    return CountScriptSigOps(subscript, SigOpCountMode::ACCURATE);
+}
diff --git a/src/script/sigops.h b/src/script/sigops.h
index 2ca2127451dd205e6eef2d63bcdf214c3bfe7ece..326008406708ae169afe362383efdc9e4edda72a 100644
--- a/src/script/sigops.h
+++ b/src/script/sigops.h
@@ -10,3 +10,5 @@ enum class SigOpCountMode {
 };
 
 uint32_t CountScriptSigOps(const CScript &script, SigOpCountMode mode);
+
+uint32_t CountScriptSigOpsP2SH(const CScript &scriptSig);
diff --git a/src/test/sigops_tests.cpp b/src/test/sigops_tests.cpp
index 30621cf01976f344e3a9b47a98c8a5836e886ec2..892195d14d95b4ab9705fb030ad7613499eea4aa 100644
--- a/src/test/sigops_tests.cpp
+++ b/src/test/sigops_tests.cpp
@@ -57,4 +57,63 @@ BOOST_AUTO_TEST_CASE(CountScriptSigOps_test) {
     }
 }
 
+BOOST_AUTO_TEST_CASE(CountScriptSigOpsP2SH_test) {
+    BOOST_CHECK_EQUAL(CountScriptSigOpsP2SH(CScript()), 0);
+    BOOST_CHECK_EQUAL(CountScriptSigOpsP2SH(CScript() << OP_CHECKSIG), 0);
+
+    std::vector<uint8_t> dummy(20);
+
+    {
+        CScript script;
+        script << OP_CHECKSIG;
+        std::vector<uint8_t> scriptBytes(script.begin(), script.end());
+        BOOST_CHECK_EQUAL(CountScriptSigOpsP2SH(CScript() << scriptBytes), 1);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << dummy << scriptBytes), 1);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << scriptBytes << dummy), 0);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << OP_CHECKSIG << scriptBytes), 0);
+    }
+
+    {
+        CScript script;
+        script << OP_CHECKSIG << OP_CHECKSIGVERIFY;
+        std::vector<uint8_t> scriptBytes(script.begin(), script.end());
+        BOOST_CHECK_EQUAL(CountScriptSigOpsP2SH(CScript() << scriptBytes), 2);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << dummy << scriptBytes), 2);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << scriptBytes << dummy), 0);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << OP_CHECKSIG << scriptBytes), 0);
+    }
+
+    {
+        CScript script;
+        script << OP_1 << dummy << dummy << OP_4 << OP_CHECKMULTISIG;
+        std::vector<uint8_t> scriptBytes(script.begin(), script.end());
+        BOOST_CHECK_EQUAL(CountScriptSigOpsP2SH(CScript() << scriptBytes), 4);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << dummy << scriptBytes), 4);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << scriptBytes << dummy), 0);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << OP_CHECKSIG << scriptBytes), 0);
+    }
+
+    {
+        CScript script;
+        script << OP_CHECKMULTISIG;
+        std::vector<uint8_t> scriptBytes(script.begin(), script.end());
+        BOOST_CHECK_EQUAL(CountScriptSigOpsP2SH(CScript() << scriptBytes), 20);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << dummy << scriptBytes), 20);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << scriptBytes << dummy), 0);
+        BOOST_CHECK_EQUAL(
+            CountScriptSigOpsP2SH(CScript() << OP_CHECKSIG << scriptBytes), 0);
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()
